[ROCm] Fix for JAX failure on ROCm due to commit ba57ae7f24743e684accef35214…

rsanthanam-amd opened the pull request:
…85a24c1235186.

When running JAX unit tests on ROCm, the following error was observed:

ImportError: /usr/local/lib/python3.9/dist-packages/jaxlib/xla_extension.so: undefined symbol: _ZN10tensorflow33tensor_float_32_execution_enabledEv

This PR fixes the error.

/cc @cheshire @hawkinsp

cheshire commented:
Is there a better target one can use? I don't believe there is currently a dependency to this one from XLA.

rsanthanam-amd commented:
I think there is a dependency in tensorflow/compiler/xla/service/gpu/gpu_layout_assignment.cc.

The commit cited in the description moved that functionality from the stream_executor but instead of specifying the new location of the target in //tensorflow/compiler/xla/service/gpu:gpu_executable, it simply removed it.

rsanthanam-amd commented:
I think this is a better target. PTAL.

rsanthanam-amd commented:
Updated to be ROCm specific.

cheshire commented:
This target is indeed better, but why is it AMD-specific? Let's just leave the dep always on?

rsanthanam-amd commented:
I think this is what happened:

The commit mentioned in the description moved the matmul_util target from the stream executor to tensorflow/core/kernels.

That target has tensorflow_float_32_utils as a dependency in both the old and new locations.

But in the old location, it applied to both rocm and cuda.  In the new location, it is CUDA only.

So after the move, the dependency is covered for CUDA but not for ROCm.

That is why this PR specifies it for ROCm only for this specific target.

The original error I encountered was an undefined symbol (tensorflow33tensor_float_32_execution_enabled) on xla_extension but presumably occurs for ROCm only.

hawkinsp commented:
@chr1sj0nes is the author of that change.

Something looks off here.
a) XLA shouldn't be depending on anything much in tensorflow/core/kernels.
b) in general targets should depend directly on what they need. Something is usually off if a top-level target is having dependencies added because an underlying target is missing them.

rsanthanam-amd committed:
Fix for JAX failure on ROCm due to commit ba57ae7f24743e684accef3521485a24c1235186.

When running JAX unit tests, the following error was observed:

ImportError: /usr/local/lib/python3.9/dist-packages/jaxlib/xla_extension.so: undefined symbol: _ZN10tensorflow33tensor_float_32_execution_enabledEv

This PR fixes the error.

rsanthanam-amd commented:
The file that has the reference is: tensorflow/compiler/xla/service/gpu/gpu_layout_assignment.h.

I have changed the PR to reflect that.

PTAL.

rsanthanam-amd commented:
@cheshire @hawkinsp 

gentle ping

cheshire commented:
Sorry I'm still confused there, why the CUDA build is not failing? If this dep is guarded with if_cuda, maybe it's better to relax that guard instead? Also do we need `if_static` there?

rsanthanam-amd commented:
I think CUDA gets the dep indirectly from //tensorflow/core/kernels:matmul_util:

```
cc_library(
    name = "matmul_util",
    srcs = if_cuda(["matmul_util.cc"]),
    hdrs = if_cuda(["matmul_util.h"]),
    deps = if_cuda([
        "@com_google_absl//absl/container:flat_hash_map",
        "//tensorflow/compiler/xla:status_macros",
        "//tensorflow/core:framework",
        "//tensorflow/core:lib",
        "//tensorflow/core/platform:tensor_float_32_hdr_lib",
        "//tensorflow/core/util:env_var",
        "//tensorflow/stream_executor/cuda:cublas_plugin",
    ]) + if_static(["//tensorflow/core/platform:tensor_float_32_utils"]),
)
```

This got moved in the PR cited in the description.

That dep has `if_static`.

cheshire commented:
But in your example tensor_float_32_utils is outside the if_cuda guard?

rsanthanam-amd commented:
I finally found the correct problem: //tensorflow/core/grappler/optimizers:generic_layout_optimizer.

In this case, I have to make a change to the ROCm TF fork, so I will close this PR without merging.

Sorry for the confusion.
